machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,243 @@
+	// +optional
+	BaseImagePullSecret *ImageSecretObjectReference `json:"baseImagePullSecret,omitempty"`
+	// machineOSImageBuilder describes which image builder will be used in each build triggered by this MachineOSConfig
+	// Currently supported type(s): JobImageBuilder
+	// +kubebuilder:validation:Required
+	ImageBuilder *MachineOSImageBuilder `json:"imageBuilder"`
+	// renderedImagePushSecret is the secret used to connect to a user registry.
+	// the final image push and pull secrets should be separate for security concerns. If the final image push secret is somehow exfiltrated,
+	// that gives someone the power to push images to the image repository. By comparison, if the final image pull secret gets exfiltrated,
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// this push secret will be used only by the MachineConfigController pod to push the image to the final destination. Not all nodes will need to push this image, most of them
+	// will only need to pull the image in order to use it.
+	// +kubebuilder:validation:Required
+	RenderedImagePushSecret ImageSecretObjectReference `json:"renderedImagePushSecret"`
+	// renderedImagePushspec describes the location of the final image.
+	// the MachineOSConfig object will use the in cluster image registry configuration.
+	// if you wish to use a mirror or any other settings specific to registries.conf, please specify those in the cluster wide registries.conf.
+	// The format of the image pushspec is:
+	// host[:port][/namespace]/name:<tag> or svc_name.namespace.svc[:port]/repository/name:<tag>
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`self.matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?(/[a-zA-Z0-9-_]{1,61})*/[a-zA-Z0-9-_.]+:[a-zA-Z0-9._-]+$') || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$')`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme. Or it must be a valid .svc followed by a port, repository, image name, and tag."
+	// +kubebuilder:validation:Required
+	RenderedImagePushspec string `json:"renderedImagePushspec"`
+	// releaseVersion is an Openshift release version which the base OS image is associated with.
+	// This field is populated from the machine-config-osimageurl configmap in the openshift-machine-config-operator namespace.
+	// It will come in the format: 4.16.0-0.nightly-2024-04-03-065948 or any valid release. The MachineOSBuilder populates this field and validates that this is a valid stream.
+	// This is used as a label in the dockerfile that builds the OS image.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
```

Comment:
What's the pattern here? Is this one of the standard formats like DNS1123Subdomain?

Reply:
Yes, will update this validation too

Reply:
It looks like this is just how all "name" type objects are validated across the MCO namespaced objects. So probably some precedent there.

If I'm grepping correctly, the other place I see this is `Host` field under route, which is talking about `DNS952` instead. Maybe all the MCO validation was inaccurate?

Reply:
Are there any tests associated to the validation that show some examples of things that do and do not pass?

Breaking down the regex, it looks like it's the same thing, except it's limiting each segment to 63 chars, where I don't think 1123 does do that 🤔 So, moving to DNS1123 would be a relaxation of the validation, but would not be incompatible I don't think